apps/aecore/src/aec_state.erl

```diff
@@ -0,0 +1,143 @@
+-export([check_chain_for_successor/2]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(SERVER, ?MODULE).
+-define(DEFAULT_CALL_TIMEOUT, infinity).
+
+-record(state, {trees  :: trees(),
+                height = ?GENESIS_HEIGHT :: height()}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+-spec get_trees() -> {ok, {height(), trees()}}.
+get_trees() ->
+    gen_server:call(?SERVER, get_trees, ?DEFAULT_CALL_TIMEOUT).
+
+-spec apply_txs(list(), height()) -> {ok, {height(), trees()}} | {{error, atom()}, trees()}.
+apply_txs(Txs, AtHeight) ->
+    gen_server:call(?SERVER, {apply_txs, {Txs, AtHeight}}, ?DEFAULT_CALL_TIMEOUT).
+
+%% API needed when external fork has more POW and we need to restart tree from common ancestor
+-spec force_trees(trees(), height()) -> {ok, {height(), trees()}}.
+force_trees(Trees, AtHeight) ->
+    gen_server:call(?SERVER, {force_trees, {Trees, AtHeight}}, ?DEFAULT_CALL_TIMEOUT).
+
+stop() ->
+    gen_server:stop(?SERVER).
+
+-spec check_chain_for_successor(trees(), height()) -> {trees(), height()}.
+check_chain_for_successor(Trees, AtHeight) ->
+    SuccessorHeight = AtHeight+1,
+    case aec_chain:get_block_by_height(SuccessorHeight) of
+        {ok, Block} ->
+            {ok, TreesUpdated} = ?MODULE:apply_txs(aec_blocks:txs(Block), SuccessorHeight),
+            check_chain_for_successor(TreesUpdated, SuccessorHeight);
+        {error, {_, _}} -> %% block_not_found, chain_too_short
+            {Trees, AtHeight}
+    end.
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+init([]) ->
+    {ok, TopBlock} = aec_chain:top_block(),
+    TopHeight = aec_blocks:height(TopBlock),
+    EmptyTrees = aec_blocks:trees(aec_block_genesis:genesis_block()),
+    {ok, TreesAtTopHeight} = setup_trees(TopHeight, EmptyTrees),
+
+    {ok, #state{trees = TreesAtTopHeight, height = TopHeight}}.
+
+handle_call(get_trees, _From, #state{trees = Trees, height = Height} = State) ->
+    {reply, {ok, {Height, Trees}}, State};
+
+handle_call({force_trees, {Trees, AtHeight}}, _From, State) ->
+    {reply, {ok, {AtHeight, Trees}}, State#state{trees = Trees, height = AtHeight}};
+
+handle_call({apply_txs, {Txs, AtHeight}}, _From,
+            #state{trees = Trees, height = CurrentHeigth} = State) ->
+    {Reply, TreesUpdated, HeightUpdated}
+        = case validate_height(CurrentHeigth, AtHeight) of
+              true ->
+                  {ok, TreesUpdated0} = apply_txs_internal(Txs, Trees, AtHeight),
+                  spawn(?MODULE, check_chain_for_successor, [TreesUpdated0, AtHeight]),
```

Comment:
May I suggest using a linked spawn, preferably `proc_lib:spawn_link/3`? I see the state service does not trap exits.

----

(I see this is an asynchronous mechanism for making such that, when a new block is inserted in the chain and the state service had the predecessor (by height), the state service attempts to consider any successor blocks that had been previously inserted in the chain without predecessor.)